### PR TITLE
RNN training optimizations

### LINKED ISSUE #1564: Improve performance of RNNS
as measured by wall time to specific training loss level. Increase batch size, gpu, and learning rate.

### PR DESCRIPTION
**Pull request by frodre**
This PR includes a new image (`prognostic_run_gpu`) built and pushed by CircleCI capable of training using our GPU nodes (`gpu-sim-pool`) on the cluster.   I upgraded the tensorflow requirement to [2.5.2](https://github.com/tensorflow/tensorflow/releases/tag/v2.5.2) to be congruent with the CUDA libraries in the prognostic run image.  Instead of forcing this update across the whole repo w/ a merge to master, I chose our paper branch for now.

Experiment configs separated into #1570 

resolves #1564

### CODE REVIEW

File: projects/microphysics/experiments/rnn-optimize/argo_gpu.yaml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      - key: "nvidia.com/gpu"
+        operator: "Exists"
+        effect: "NoSchedule"
+    - name: score-model
+      inputs:
+        parameters:
+          - name: training-config
+          - {name: flags, value: " "}
+      container:
+        image: "us.gcr.io/vcm-ml/prognostic_run{{workflow.parameters.tag}}"
```
Comment (frodre) [inline projects/microphysics/experiments/rnn-optimize/argo_gpu.yaml:101]: Should have some conditional GPU logic and merge this into the training argo (or add a new template to the training argo)

File: Makefile [hunk 2]
```diff
@@ -26,7 +26,14 @@ push_images: $(addprefix push_image_, $(IMAGES))
 build_image_prognostic_run:
 	tools/docker_build_cached.sh us.gcr.io/vcm-ml/prognostic_run:$(CACHE_TAG) \
 		-f docker/prognostic_run/Dockerfile -t $(REGISTRY)/prognostic_run:$(VERSION) \
-		--target prognostic-run .
+		--target prognostic-run \
+		--build-arg BASE_IMAGE=ubuntu:20.04 .
+
+build_image_prognostic_run_gpu:
```
Comment (frodre) [inline Makefile:32]: Opted to build from an CUDA/ubuntu base image since it seemed easy, but this image is big (around 7-8 GB) w/ the CUDA libraries.

File: constraints.txt [hunk 3]
```diff
@@ -29,6 +29,7 @@ blinker==1.4
 bokeh==2.2.3
 bump2version==1.0.0
 cachecontrol==0.12.6
+cached-property==1.5.2
```
Comment (frodre) [inline constraints.txt:32]: BSD License

File: constraints.txt [hunk 4]
```diff
@@ -127,6 +128,7 @@ jupyterlab-pygments==0.1.2
 jupyterlab-server==2.3.0
 jupyterlab-widgets==1.0.0
 jupyterlab==3.0.9
+keras-nightly==2.5.0.dev2021032900
```
Comment (frodre) [inline constraints.txt:131]: Apache Software License (Apache 2.0)

File: constraints.txt [hunk 5]
```diff
@@ -245,11 +247,12 @@ sphinxcontrib-qthelp==1.0.3
 sphinxcontrib-serializinghtml==1.1.4
 streamlit==0.82.0
 subprocess32==3.5.4
+tensorboard-data-server==0.6.1
```
Comment (frodre) [inline constraints.txt:250]: Apache Software License (Apache 2.0)

File: docker/prognostic_run/scripts/install_esmf.sh [hunk 6]
```diff
@@ -14,7 +14,7 @@ export ESMF_NETCDF_INCLUDE=/usr/include
 export ESMF_NETCDF_LIBS="-lnetcdf -lnetcdff"
 export ESMF_BOPT=O3
 
-git clone -b ESMF_8_0_0 --depth 1 https://git.code.sf.net/p/esmf/esmf $ESMF_DIR
+git clone -b ESMF_8_0_0 --depth 1 https://github.com/esmf-org/esmf.git $ESMF_DIR
```
Comment (frodre) [inline docker/prognostic_run/scripts/install_esmf.sh:17]: The old URL was 404, wasn't sure where the previous repos was

File: external/fv3fit/setup.py [hunk 7]
```diff
@@ -11,7 +11,7 @@
     "scikit-learn>=0.22",
     "fsspec>=0.6.2",
     "pyyaml>=5.1.2",
-    "tensorflow>=2.2.0",
+    "tensorflow==2.5.2",
```
Comment (frodre) [inline external/fv3fit/setup.py:14]: Needed to match CUDA library compatibility in the chosen image.

File: projects/microphysics/scripts/prognostic_run.py [hunk 8]
```diff
@@ -24,6 +24,8 @@ def get_env(args):
     env["SAVE_ZARR"] = args.save_zarr
     env["SAVE_TFRECORD"] = args.save_tfrecord
     env["SAVE_NC"] = args.save_nc
+
+    env = {k: str(v) for k, v in env.items()}
```
Comment (frodre) [inline projects/microphysics/scripts/prognostic_run.py:28]: I was getting an error about a boolean environment variable where it was expecting a string when running the prognostic evaluation.

File: .circleci/config.yml [hunk 9]
```diff
@@ -93,6 +93,7 @@ jobs:
         type: enum
         enum:
           - prognostic_run
+          - prognostic_run_gpu
```
Comment (frodre) [inline .circleci/config.yml:96]: New image for GPU training/prognostic runs.

File: external/fv3fit/setup.py (hunk 7 above)
Comment (nbren12) [reply external/fv3fit/setup.py:14]: Maybe add a comment saying this?

File: projects/microphysics/scripts/prognostic_run.py (hunk 8 above)
Comment (nbren12) [reply projects/microphysics/scripts/prognostic_run.py:28]: same. I had to fix this in my pr too. Would be nice to keep this commit separate from the other changes.

**Review by nbren12 — APPROVED** [review]
Thanks! Just a couple requests
1. can you add an rnn training config with the optimal parameters?
2. Merge this as rebase+merge. The commit history is a little long but nicely documents this set of experiments. It also has some bug fixes that would be nice to go in as separate commits.

**Comment by nbren12** [comment]
I guess I was confused and thought this was going into master. I opened #1572 to do this.
